Add return_scaled to RecordingExtractor get_traces()

luiztauffer opened the pull request:
fix #540

samuelgarcia commented:
Hi again.
it is not the only kwargs. You should also do something like  

```python
    def __init__(self, filename: PathType, nsx_to_load: Optional[int] = None,
                 block_index: Optional[float] = None, seg_index: Optional[float] = None, **kwargs):
        super().__init__(filename=filename, nsx_to_load=nsx_to_load, 
                         block_index=block_index, seg_index=seg_inde, **kwargsx)
```

Normaly this already done automatically by the NeoBaseRecordingExtractor.
This implementation only make it more explicit.

luiztauffer commented:
how about now?

luiztauffer commented:
Somehow related to this, the `get_traces` method returns the traces already scaled to `uV`, when it would be interesting in some cases to have them as `int`and then have the gain applied whenever needed. That helps us save a lot of space.

I added an argument to `NeoBaseRecordingExtractor.get_traces(..., return_scaled=True)` that gives a choice for that, defaults to True (current behavior)

Also, I added channel properties `gain` and `name` at __init__

samuelgarcia commented:
This gain is a good idea I add this in mind from a long time.

I wanted to do it at base level not only for neo wrapper.

We already have this option in neo, it is super usefull for spike sorting and saving memory bandwith.

OK for this `return_scaled` but it may change a bit in a futur "big refactoring" I have started very recently.

alejoe91 commented:
@luiztauffer still failing but not for your fault :P 

I know why it's failing and I'll fix it over the weekend. It's due to the decorator that we have for handling input arguments to `get_traces`, which somehow it's messing things up.

luiztauffer commented:
> @luiztauffer still failing but not for your fault :P
> 
> I know why it's failing and I'll fix it over the weekend. It's due to the decorator that we have for handling input arguments to `get_traces`, which somehow it's messing things up.

thanks @alejoe91 
I just added the new arg `return_scaled`, should the decorator be aware of that?

The `get_traces` method works fine in my code though...

alejoe91 commented:
It should deal with it, but now it's base on the base class signature. I'll update it tomorrow to handle all base args plus class-specific kwargs

alejoe91 commented:
@luiztauffer @samuelgarcia @bendichter 

This required a little more extensive fix than expected. I added `return_scaled` in the base `RecordingExtractor` method and propagated to all extractors. I also improved the way decorators handle the arguments and kwargs. 

Let's check that everything looks correct!

Note that this also needed an update to `spiketoolkit`, PR coming soon.

samuelgarcia commented on spikeextractors/extractors/bindatrecordingextractor/bindatrecordingextractor.py:
```diff
@@ -77,7 +77,7 @@ def get_traces(self, channel_ids=None, start_frame=None, end_frame=None):
             recordings = recordings.astype('float32') - 2**(exp - 1)
-        if self._gain is not None:
+        if self._gain is not None and return_scaled:
             recordings = recordings * self._gain
```
I gain is None should and return_scaled=True should raise an error

alejoe91 replied:
I think we should do the following. If gain is None in the constructor, we set it to one. Then we multiply by whatever gain we have. I use the bindat a lot (cache) to save filtered data already converted to uV. In that case the scale should be one, bit I think it should be so by default.

alejoe91 replied:
Fixing BinDat on another PR

samuelgarcia commented on spikeextractors/extractors/mea1kextractors/mea1kextractors.py:
```diff
@@ -181,17 +181,18 @@ def get_sampling_frequency(self):
-            return (self._signals[np.array(channel_ids), start_frame:end_frame] * self._lsb).astype('float32')
+            signals = self._signals[np.array(channel_ids), start_frame:end_frame]
+        if return_scaled:
+            return (signals * self._lsb).astype('float32')
```
```
signals = signals.astype('float32')
signals *=  * self._lsb
```

would avoid a copy I think

alejoe91 replied:
fixed

samuelgarcia commented on spikeextractors/extractors/neoextractors/neobaseextractor.py:
```diff
@@ -78,28 +78,42 @@ def __init__(self, block_index=None, seg_index=None, **kargs):
-        scaled_traces = scaled_traces.transpose()
+            # fortunatly neo works with (samples, channels) strides
+            # so transpose to spieextractors wolrd
+            scaled_traces = scaled_traces.transpose()
```
scaled_traces = scaled_traces.transpose() should not be in "if return_scaled:"

alejoe91 replied:
fixed

alejoe91 commented:
@CodyCBakerPhD merging into the `return_scaled` branch. Let's finish it up on that!
